Sources/Core/GTMSessionFetcherService.m

```diff
@@ -315,18 +316,14 @@ - (NSURLSession *)sessionForFetcherCreation {
     // fetcher, and when the fetcher attempts to create its NSURLSessionTask.
     [_delegateDispatcher startSessionUsage];
 
-    NSURLSession *session = _delegateDispatcher.session;
-    if (session) {
-      // The calling fetcher will receive a preexisting session, so
-      // we can allow other fetchers to create a session.
-      dispatch_semaphore_signal(_sessionCreationSemaphore);
-    } else {
-      // No existing session was obtained, so the calling fetcher will create the session;
-      // it *must* invoke fetcherDidCreateSession: to signal the dispatcher's semaphore after
-      // the session has been created (or fails to be created) to avoid a hang.
+    session = _delegateDispatcher.session;
+    if (!session) {
+      session = creationBlock(_delegateDispatcher);
```

Comment:
Try/catch on this to avoid going to a bad state?

Reply:
Or maybe just a try/finally?  Or are you comfortable in it should never throw?

Reply:
Was thinking it should be fine, but for safety's sake yeah it should probably have try/finally to ensure the lock is released.